upgrade-os

```diff
@@ -0,0 +1,158 @@
+#!/usr/bin/env bash
+set -e
+
+export LOG_FILE=/tmp/update.$(date +"%Y%m%d").log
+
+SCRIPTPATH="$( cd "$(dirname "$0")" ; pwd -P )"
+UBUNTU_VERSION=$(lsb_release -rs)
+
+rm -f ${LOG_FILE}
+
+decho () {
+  echo `date +"%H:%M:%S"` $1
+  echo `date +"%H:%M:%S"` $1 >> ${LOG_FILE}
+}
+
+checkDpkgLock () {
+  while sudo fuser /var/lib/dpkg/lock /var/lib/apt/lists/lock /var/cache/apt/archives/lock >/dev/null 2>&1
+  do
+    echo 'Operating system is using dpkg/apt locks, waiting for their release...'
+    sleep 5
+  done
+}
+
+checkIfRebootNeeded () {
+  if [ -f /var/run/reboot-required ]
+  then
+    echo -e "\033[0;33mThe server will now reboot, resuming the update process directly after.\033[0m"
+    echo -e "\033[0;33mDuring this time, paired machines will not be online.\033[0m"
+    echo
+    echo -e "Please wait 10 seconds before logging in, and consult the update process with the following command:"
+    echo -e "\033[0;33mcheck-upgrade-status\033[0m"
+    sudo shutdown -r now
+  fi
+}
+
+checkIfCanContinue () {
+  checkIfRebootNeeded
+  checkDpkgLock
+}
+
+cat <<'FIG'
+ _
+| | __ _ _ __ ___   __ _ ___ ___ _   _       ___  ___ _ ____   _____ _ __
+| |/ _` | '_ ` _ \ / _` / __/ __| | | |_____/ __|/ _ \ '__\ \ / / _ \ '__|
+| | (_| | | | | | | (_| \__ \__ \ |_| |_____\__ \  __/ |   \ V /  __/ |
+|_|\__,_|_| |_| |_|\__,_|___/___/\__,_|     |___/\___|_|    \_/ \___|_|
+FIG
+
+echo -e "\nStarting \033[1mlamassu-server\033[0m OS update. This will take a while...\n"
+
+if [ "$(whoami)" != "root" ]; then
+  echo -e "This script has to be run as \033[1mroot\033[0m user"
+  exit 3
+fi
+
+# Use a lock file so failed scripts cannot be imediately retried
+# If not the backup created on this script would be replaced
+if ! mkdir /var/lock/lamassu-update; then
+  echo "Script is locked because of a failure." >&2
+  exit 1
+fi
+
+if [ -f /usr/bin/check-upgrade-status ]; then
+  rm -r /usr/bin/check-upgrade-status
+fi
+
+echo "tail -n 10 ~/upgrade.log" > /usr/bin/check-upgrade-status
+sudo chmod +x /usr/bin/check-upgrade-status
+
+if [[ "$UBUNTU_VERSION" == "16.04" ]]; then
+  echo "Detected Ubuntu version: 16.04. Updating to Ubuntu version: 18.04..."
+  echo
+  echo -e "\033[0;33mNOTE: Some of the system's dependencies might require more reboots than expected during this process.\033[0m"
+  echo -e "\033[0;33mThe operating system upgrade process should take around 30 minutes.\033[0m"
+  echo -e "\033[0;33mYou can check the upgrade process after the first reboot, by running the following command:\033[0m"
+  echo
+  echo -e "\033[0;33mcheck-upgrade-status\033[0m"
```

Comment:
No need to note about the dependencies requiring multiple reboots. We'll prompt to stay out of the server for 30 minutes anyway.

```
  echo -e "\033[0;33mThe server operating system upgrade should take around 30 minutes total.\033[0m"
  echo
  echo -e "\033[0;33mDuring this time, your paired machines will be offline.\033[0m"
  echo
  echo -e "\033[0;33mPlease pay attention in a few minutes to the end of the following output for next steps to take.\033[0m"
```